### PR TITLE
DATAMONGO-1245 - Add support for Query By Example.

### PR DESCRIPTION
**Pull request by christophstrobl**
We now support querying documents by providing a sample of the given object holding compare values.

```java
Example<Person> example = newExampleOf(new Person("dAVe", "Matth"))
    .ignoreNullValues()
    .matchStringsStartingWith()
    .matchStringsWithIgnoreCase()
    .get();

Page<Person> page = repository.findAllByExample(example, new PageRequest(0, 10));
```

For the sake of partial matching we flatten out nested structures (depending on the configured `NullHandler`) so we can create different queries for matching like:

```javascript
{ _id : 1, nested : { value : "conflux" } }
{ _id : 1, nested.value : { "conflux" } }
```

This is useful when you want so search using a only partially filled nested document.

Usage of `StringMatcher` allows usage of `$regex` operator.
`DBRefs` and geo structures such as `Point` or `GeoJsonPoint` are converted to their according structure.

----

Requires: spring-projects/spring-data-commons#153 (DATACMNS-810)

### CODE REVIEW

**Comment by mp911de** [comment]
@gregturn Can you have a look on the docs pls?

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+* Querying your data store with a set of static or dynamic constraints
+* Frequent refactoring of the entities without worrying about breaking existing queries
+* Works independently from the data store API
+
+**Limitations**
+
+* Query predicates are combined using the `AND` keyword
+* No support for nested/grouped property constraints like `firstname = ?0 or (firstname = ?1 and lastname = ?2)`
+* Limited to starts/contains/ends/regex matching for strings and exact matching for other property types
+
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: Instead of repeating "Limited to...", say "Only supports..."

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  private String firstname;
+  private String lastname;
+  private Address address;
+
+  // … getters and setters omitted
+}
+----
+====
+
+We have a quite simple entity here that is mapped to the data store. You can use this entity to create an Example specification. By default, fields having `null` values are ignored, and strings are matched using the store specific defaults. Examples can be built by either using the `exampleOf` factory method or by using the <<query.by.example.builder,Example builder>>. Once the `Example` is constructed it becomes immutable.
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: Pick "we" or "you" and stick with it for consistency.

"This is a simple domain object. You can use it to create an Example specification."

I hear "entity" and think "JPA", so I lean towards domain object for general purposes.

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+.The first Example specification
+====
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: . First example specification

For headers, I try to avoid "extra" words like "the". Keep it slim.

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+----
+<1> Create a new instance of the entity
+<2> Set the properties to query
+<3> Create an `Example`
+====
+
+
+NOTE: Property names of the sample object must correlate with the property names of the queried entity.
+
+.Query by Example using the MongoTemplate
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: .Query by Example using MongoTemplate

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+MongoTemplate template;
+
+public List<Person> findPeople(Person sampleObject) {
+  return template.findByExample(Example.exampleOf(person));
+}
+----
+====
+
+The `findByExample` method accepts either the sample object or an `Example` object to query the data store. Spring Data uses the `Example` to create and execute a query.
+
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: Feels like these two sentences are saying the same thing. I would collapse into a single sentence.

"MongoTemplate.findByExample() accepts either the sample object or an `Example` object create and execute a query."

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+.Query by Example using the Repositories
+====
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: .Query by Example using a Repository

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+  // … more functionality omitted.
+}
+----
+====
+
+[[query.by.example.builder]]
+== Example builder
+
+Examples are not limited to default settings. You can specify own defaults for string matching, null handling and property-specific settings using the example builder.
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: ...You can customize the settings for string matching...

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            newPropertySpecifier("firstname").matchString(StringMatcher.CONTAINING).get())
+    .withPropertySpecifier(
+            newPropertySpecifier("lastname").matchStringsWithIgnoreCase().get())
+    .withPropertySpecifier(
+            newPropertySpecifier("address.city").matchStringStartingWith().get())
+    .get();
+----
+====
+
+Property specifier accepts property names and property paths separated by dots that are contained within the sample object. A `PropertySpecifier` allows setting string matching options, case-sensitivity,
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: Property specifier accepts property names (e.g. "firstname" and "lastname"). You can navigate by chaining properties together with dots ("address.city"). You can tune it with matching options, case sensitivity...

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+[cols="1,2", options="header"]
+.Supported string matching options of `StringMatcher`
+|===
```
Comment (gregturn) [inline src/main/asciidoc/reference/query-by-example.adoc]: .`StringMatcher` options

(It's implicit that this list is supported.)

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoOperations.java [hunk 10]
```diff
@@ -629,6 +631,36 @@
...
+	 * configured otherwise, an instance of MappingMongoConverter will be used.
+	 * <p/>
+	 * The query is constructed from the {@code probe}.
+	 *
+	 * @param probe the probe to be used for the Example, must not be {@literal null}.
+	 * @param <S>
+	 * @param <T>
+	 * @return the List of converted objects
+	 */
+	<S extends T, T> List<T> findByExample(S probe);
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoOperations.java]: I get compile errors here due to `S` referring to `T` defined later, inverting the declaration should fix that. Same below.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoOperations.java [hunk 11]
```diff
@@ -56,6 +57,7 @@
  * @author Chuong Ngo
  * @author Christoph Strobl
  * @author Thomas Darimont
+ * @author Mark Paluch
  */
 public interface MongoOperations {
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoOperations.java]: Please revert changes on `MongoOperations` and `MongoTemplate` completely clients can use the `Criteria`-API to work with examples and use existing methods taking a `Query` instance.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java [hunk 12]
```diff
@@ -117,6 +117,10 @@ public Point convert(DBObject source) {
 
 			Assert.isTrue(source.keySet().size() == 2, "Source must contain 2 elements");
 
+			if (source.containsField("type")) {
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java]: This change seems to be unrelated to the PR. @christophstrobl — can you comment?

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MongoExampleMapper.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+		}
+
+		ExampleSpecAccessor exampleSpecAccessor = new ExampleSpecAccessor(example.getExampleSpec());
+
+		applyPropertySpecs("", reference, example.getProbeType(), exampleSpecAccessor);
+
+		if (exampleSpecAccessor.isTyped()) {
+			Set<Class<?>> restrictedTypes = new HashSet<Class<?>>();
+			restrictedTypes.add(example.getResultType());
+			this.converter.getTypeMapper().writeTypeRestrictions(reference, restrictedTypes);
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MongoExampleMapper.java]: Maybe rather `Collections.singletonSet(…)`?

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MongoExampleMapper.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+		MongoPersistentEntity<?> entity = mappingContext.getPersistentEntity(probeType);
+
+		Iterator<String> parts = Arrays.asList(path.split("\\.")).iterator();
+
+		final Stack<MongoPersistentProperty> stack = new Stack<MongoPersistentProperty>();
+
+		List<String> resultParts = new ArrayList<String>();
+
+		while (parts.hasNext()) {
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MongoExampleMapper.java]: I'd love to see the nesting reduced here. Can we introduce eager returns/continues?

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java [hunk 15]
```diff
@@ -254,8 +261,8 @@ protected DBObject getMappedKeyword(Field property, Keyword keyword) {
 		boolean needsAssociationConversion = property.isAssociation() && !keyword.isExists();
 		Object value = keyword.getValue();
 
-		Object convertedValue = needsAssociationConversion ? convertAssociation(value, property)
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java]: Seems to be formatting only.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java [hunk 16]
```diff
@@ -477,8 +484,8 @@ public Object convertId(Object id) {
 		}
 
 		try {
-			return conversionService.canConvert(id.getClass(), ObjectId.class) ? conversionService.convert(id, ObjectId.class)
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java]: Seems to be formatting only.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/SerializationUtils.java [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012 the original author or authors.
+ * Copyright 2012-2015 the original author or authors.
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/SerializationUtils.java]: 2016.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/SerializationUtils.java [hunk 18]
```diff
@@ -37,6 +41,68 @@ private SerializationUtils() {
...
+	 *   nested.value : "conflux"
+	 * }
+	 * </code>
+	 * </pre>
+	 * 
+	 * @param source can be {@literal null}.
+	 * @return {@link Collections#emptyMap()} when source is {@literal null}
+	 * @since 1.8
+	 */
+	public static Map<String, Object> flatMap(DBObject source) {
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/SerializationUtils.java]: Maybe rather `flattenMap(…)` to avoid confusion with a functional flat-map?

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java (hunk 12 above)
Comment (christophstrobl) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java]: true - I'll open an issue for that one.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/MongoRepository.java [hunk 19]
```diff
@@ -18,43 +18,50 @@
...
  * Mongo specific {@link org.springframework.data.repository.Repository} interface.
- * 
+ *
  * @author Oliver Gierke
  * @author Christoph Strobl
  * @author Thomas Darimont
+ * @author Mark Paluch
  */
 @NoRepositoryBean
-public interface MongoRepository<T, ID extends Serializable> extends PagingAndSortingRepository<T, ID> {
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/MongoRepository.java]: Some formatting related changes. Adding the `@Override`s is not a bad idea, but we should keep them in a separate commit or simply postpone them.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/ConvertingParameterAccessor.java [hunk 20]
```diff
@@ -132,6 +133,11 @@ public TextCriteria getFullText() {
 		return delegate.getFullText();
 	}
 
+	@Override
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/ConvertingParameterAccessor.java]: Shouldn't be necessary anymore.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParameterAccessor.java [hunk 21]
```diff
@@ -60,4 +61,12 @@
 	 * @since 1.8
 	 */
 	Object[] getValues();
+
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParameterAccessor.java]: Shouldn't be necessary anymore.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParameters.java [hunk 22]
```diff
@@ -20,6 +20,7 @@
 import java.util.List;
 
 import org.springframework.core.MethodParameter;
+import org.springframework.data.domain.Example;
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParameters.java]: Shouldn't be necessary anymore (any changes in this class).

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParametersParameterAccessor.java [hunk 23]
```diff
@@ -18,6 +18,7 @@
 import java.util.Arrays;
 import java.util.List;
 
+import org.springframework.data.domain.Example;
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoParametersParameterAccessor.java]: Shouldn't be necessary anymore.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoQueryCreator.java [hunk 24]
```diff
@@ -151,7 +151,20 @@ protected Criteria or(Criteria base, Criteria criteria) {
 	@Override
 	protected Query complete(Criteria criteria, Sort sort) {
 
-		Query query = (criteria == null ? new Query() : new Query(criteria)).with(sort);
+		Criteria toUse = null;
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoQueryCreator.java]: These changes shouldn't be necessary anymore (deletion below still applies).

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/support/SimpleMongoRepository.java [hunk 25]
```diff
@@ -68,6 +71,7 @@ public SimpleMongoRepository(MongoEntityInformation<T, ID> metadata, MongoOperat
 	 * (non-Javadoc)
 	 * @see org.springframework.data.repository.CrudRepository#save(java.lang.Object)
 	 */
+	@Override
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/support/SimpleMongoRepository.java]: Let's fix the `@Overrides` in another commit.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/support/SimpleMongoRepository.java [hunk 26]
```diff
@@ -259,6 +282,92 @@ public void deleteAll() {
...
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.mongodb.repository.MongoRepository#findAllByExample(org.springframework.data.domain.Example)
+	 */
+	@Override
+	public <S extends T> List<S> findAll(Example<S> example) {
+		return findAll(example, (Sort) null);
+	}
+
+	@Override
```
Comment (olivergierke) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/support/SimpleMongoRepository.java]: Let's add the non-Javadoc comments for consistency reasons.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java [hunk 27]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012 the original author or authors.
+ * Copyright 2012-2015 the original author or authors.
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java]: 2016.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java [hunk 28]
```diff
@@ -20,18 +20,24 @@
 import static org.springframework.data.mongodb.core.query.SerializationUtils.*;
 
 import java.util.Arrays;
+import java.util.Map;
 
 import org.hamcrest.Matcher;
+import org.hamcrest.collection.IsMapContaining;
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java]: Rather use static imports for `Matchers`.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java [hunk 29]
```diff
@@ -60,6 +66,74 @@ public void writesCollection() {
...
+	public void flatMapSouldAppendCommandsCorrectly() {
+
+		DBObject dbo = new BasicDBObjectBuilder().add("_id", 1)
+				.add("nested", new BasicDBObjectBuilder().add("$regex", "^conflux$").add("$options", "i").get()).get();
+
+		Map<String, Object> map = flatMap(dbo);
+
+		assertThat(map, IsMapContaining.<String, Object> hasEntry("_id", 1));
+		assertThat(map.get("nested"), notNullValue());
+		assertThat(((Map<String, Object>) map.get("nested")).get("$regex"), Is.<Object> is("^conflux$"));
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/SerializationUtilsUnitTests.java]: There is `DBObjectTestUtils` that might make things easier here.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/AbstractPersonRepositoryIntegrationTests.java [hunk 30]
```diff
@@ -175,8 +181,8 @@ public void findsPagedPersons() throws Exception {
 	@Test
 	public void executesPagedFinderCorrectly() throws Exception {
 
-		Page<Person> page = repository.findByLastnameLike("*a*",
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/AbstractPersonRepositoryIntegrationTests.java]: Let's remove all the reformattings here.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/Contact.java [hunk 31]
```diff
@@ -27,8 +27,7 @@
 @Document
 public abstract class Contact {
 
-	@Id
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/Contact.java]: Let's remove the reformatting here.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/ContactRepositoryIntegrationTests.java [hunk 32]
```diff
@@ -43,4 +53,37 @@ public void readsAndWritesContactCorrectly() {
 
 		assertTrue(repository.findOne(result.getId().toString()) instanceof Person);
 	}
+
+	@Test
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/ContactRepositoryIntegrationTests.java]: Ticket references.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/PersonRepository.java [hunk 33]
```diff
@@ -334,22 +334,23 @@
 	 */
 	@Query("{ firstname : { $in : ?0 }}")
 	Stream<Person> findByCustomQueryWithStreamingCursorByFirstnames(List<String> firstnames);
-
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/PersonRepository.java]: Let's remove all reformattings here.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/query/StubParameterAccessor.java [hunk 34]
```diff
@@ -147,4 +148,13 @@ public TextCriteria getFullText() {
 	public Class<?> getDynamicProjection() {
 		return null;
 	}
+
```
Comment (olivergierke) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/repository/query/StubParameterAccessor.java]: Shouldn't be necessary anymore.

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 35]
```diff
@@ -0,0 +1,90 @@
+[[query.by.example.execution]]
+=== Executing Example
+
+In Spring Data MongoDB you can use Query by Example with the `MongoTemplate` and with Repositories.
```
Comment (olivergierke) [inline src/main/asciidoc/reference/query-by-example.adoc]: The section on how to use the tempalte API can be removed.

File: src/main/asciidoc/reference/query-by-example.adoc [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+====
+
+`MongoTemplate.findByExample` accepts either the probe or an `Example` object to create and execute a query. An `Example` containing an untyped `ExampleSpec` uses the probe type to determine the collection name and the result type. Typed `ExampleSpec` use their type to determine collection name and result type.
+
+
+.Query by Example using a Repository
+====
+[source, java]
+----
+public interface PersonRepository extends MongoRepository<Person, String> {
```
Comment (olivergierke) [inline src/main/asciidoc/reference/query-by-example.adoc]: Let's extend the `QBEExecuter` interface rather than `MongoRepository`.

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java (hunk 12 above)
Comment (christophstrobl) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/GeoConverters.java]: created [DATAMONGO-1390](https://jira.spring.io/browse/DATAMONGO-1390).
